src/olympia/blocklist/mlbf.py

```diff
@@ -287,23 +309,47 @@ def generate_and_write_stash(self, previous_mlbf: 'MLBF' = None):
         in the "unblocked" stash (like for hard blocked items) as this would
         result in the version being in both blocked and unblocked stashes.
         """
-        diffs = self.generate_diffs(previous_mlbf)
-        blocked_added, blocked_removed, _ = diffs[BlockType.BLOCKED]
-        stash_json = {
-            'blocked': blocked_added,
-            'unblocked': blocked_removed,
+        # Map block types to hard coded stash keys for compatibility
+        # with the expected keys in remote settings.
+        STASH_KEYS = {
+            BlockType.BLOCKED: 'blocked',
+            BlockType.SOFT_BLOCKED: 'softblocked',
         }
+        UNBLOCKED_STASH_KEY = 'unblocked'
 
-        if waffle.switch_is_active('enable-soft-blocking'):
-            soft_blocked_added, soft_blocked_removed, _ = diffs[BlockType.SOFT_BLOCKED]
-            stash_json['softblocked'] = soft_blocked_added
-            stash_json['unblocked'] = [
-                unblocked
-                for unblocked in (blocked_removed + soft_blocked_removed)
-                if unblocked not in blocked_added
-            ]
+        # Base stash includes all of the expected keys from STASH_KEYS + unblocked
+        stash_json = {key: [] for key in [UNBLOCKED_STASH_KEY, *STASH_KEYS.values()]}
+
+        diffs = self.generate_diffs(previous_mlbf)
+        soft_block_enabled = waffle.switch_is_active('enable-soft-blocking')
+        set_blocked_any_type = set()
+        # Build the stash json from the diff of each block type
+        for block_type in BlockType:
+            skip_no_changes = not self.should_upload_stash(block_type, previous_mlbf)
+            skip_uploading_filter = self.should_upload_filter(block_type, previous_mlbf)
+            skip_soft_block = (
+                block_type == BlockType.SOFT_BLOCKED and not soft_block_enabled
+            )
 
-        # write stash
+            # Skip the stash if any of the conditions are met
+            if any([skip_no_changes, skip_uploading_filter, skip_soft_block]):
+                continue
+
+            added, removed, _ = diffs[block_type]
+            # Items added to the block type are included in that block list
+            # using the hard coded stash key
+            stash_json[STASH_KEYS[block_type]] = added
+            # TODO: add a better test for this.
+            set_blocked_any_type.update(added)
+            # Items removed from that block type are included in the unblocked list
+            for item in removed:
+                # Items already on the hard block list should be excluded from the
+                # unblocked list to prevent double counting of a version that has moved
+                # from soft to hard blocked.
+                if item not in set_blocked_any_type:
+                    stash_json[UNBLOCKED_STASH_KEY].append(item)
```

Comment:
It's probably more of a note for myself than anything else: we should double check what stash we get when we have a block that goes from hard to soft AND enough other changes to make a new soft filter. In this PR, we should probably not account for "old" clients with the "super-mega-unified stash" approach documented in https://docs.google.com/document/d/1JmGUPD9fZ8ri_xdF1zBwOXsQHfBX-51fV2HYbBDsFQg/edit?usp=sharing

Reply:
I'll add a test for this tomorrow.

Reply:
I added the test and, perhaps interestingly, We might be thinking about this in a less than accurate way. When we run the cron, and a block has moved from hard to soft, and there are still (or now) enough soft blocks for a new filter, what happens is:

- if soft blocking is enabled, then we get a new soft blocked filter which will include the soft blocked items.
- else we don't get a soft block filter

What happens next depends on the number of hard blocks at the time of generation:

- if there are enough hard blocks for a filter, we get 2 filters
- else if there are enough hard blocks for a stash, we get a stash with those hard blocks and an empty soft block
- else we get no stash because there are no hard blocks and soft blocks are either ignored due to waffle switch or included in the filter instead due to the number of changes

Because there are (at the time of generation) enough soft blocks for a new filter we get the filter or not depending on the state of the waffle switch. The expected stash depends on the number of hard blocks and (partially) on the state of the waffle switch.